### PR TITLE
Migrate to purescript 0.12

### PR DESCRIPTION
**Pull request by np**

### CODE REVIEW

File: src/Thermite.purs [hunk 1]
```diff
@@ -71,33 +75,27 @@ import ReactDOM (render)
...
 -- | A type synonym for an event handler which can be used to construct
 -- | `purescript-react`'s event attributes.
-type EventHandler =
-  forall eff refs.
-    Eff ( props :: React.ReactProps
-        , state :: React.ReactState React.ReadWrite
-        , refs :: React.ReactRefs refs
-        | eff
-        ) Unit
+type EventHandler = Effect Unit
```
Comment (paf31) [inline src/Thermite.purs]: I think we can remove this now.

File: src/Thermite.purs [hunk 2]
```diff
@@ -224,78 +231,77 @@ createReactSpec = createReactSpec' div'
...
+  -> Effect Unit
 defaultMain spec initialState props = void do
-  let component = createClass spec initialState
-  document <- DOM.window >>= DOM.document
+  let component = createClass "DefaultMain" spec initialState
+  window <- DOM.window
+  document <- DOM.document window
   container <- DOM.body document
-  traverse_ (render (createFactory component props) <<< DOM.htmlElementToElement) container
+  traverse_ (render (React.unsafeCreateLeafElement component props)) (DOM.toElement <$> container)
```
Comment (paf31) [inline src/Thermite.purs]: Why move the `DOM.toElement` out? It adds an extra traversal.

File: yarn.lock [hunk 3]
```diff
@@ -0,0 +1,2760 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (paf31) [inline yarn.lock]: Please can you add this file to `.gitignore`?

File: src/Thermite.purs [hunk 4]
```diff
@@ -38,30 +39,33 @@ module Thermite
...
 import React.DOM (div')
 import ReactDOM (render)
+import Web.HTML (window) as DOM
+import Web.HTML.HTMLDocument (body) as DOM
+import Web.HTML.HTMLElement (toElement) as DOM
+import Web.HTML.Window (document) as DOM
+
+import Record.Unsafe (unsafeDelete)
+
+type StateCoTransformer state a =
```
Comment (paf31) [inline src/Thermite.purs]: Can you please remove the `a` on both sides here and rename this to something shorter? Maybe even `Thermite`?

Comment (paf31) [reply src/Thermite.purs]: Can you please also add a short comment?

File: src/Thermite.purs [hunk 5]
```diff
@@ -174,48 +172,57 @@ _render = lens (\(Spec s) -> s.render) (\(Spec s) r -> Spec (s { render = r }))
...
          , render:              \k p s   -> spec1.render k p s <> spec2.render k p s
          }
 
-instance monoidSpec :: Monoid (Spec eff state props action) where
+instance monoidSpec :: Monoid (Spec state props action) where
   mempty = simpleSpec (\_ _ _ -> pure unit)
                       (\_ _ _ _ -> [])
 
+type ReactSpecSimple props state
+   = React.ReactThis {children :: Children | props} (Record state)
```
Comment (paf31) [inline src/Thermite.purs]: Can you please use the same formatting of record types here?

Comment (paf31) [reply src/Thermite.purs]: Also, why is this necessary? Doesn't `purescript-react` have a type for specs?

File: src/Thermite.purs [hunk 6]
```diff
@@ -174,48 +172,57 @@ _render = lens (\(Spec s) -> s.render) (\(Spec s) r -> Spec (s { render = r }))
...
-     , dispatcher :: React.ReactThis props state -> action -> EventHandler
+  :: forall state props action
+   . Spec (Record state) (Record props) action
+  -> Record state
+  -> { spec :: ReactSpecSimple props state
+     , dispatcher :: React.ReactThis {children :: Children | props} (Record state) -> action -> EventHandler
      }
 createReactSpec = createReactSpec' div'
 
+noChildren :: forall props. {children :: Children | props} -> Record props
```
Comment (paf31) [inline src/Thermite.purs]: Why is this needed?

**Comment by paf31** [comment]
Thanks for the PR!

File: bower.json [hunk 7]
```diff
@@ -16,12 +16,18 @@
...
+    "purescript-web-html": "^1.0.0",
+    "purescript-freet": "^4.0.0",
+    "purescript-profunctor-lenses": "^4.0.0",
+    "purescript-react-dom": "^6.0.0",
+      "purescript-react": "^6.1.0",
+      "purescript-typelevel-prelude" : "^3.0.0",
+      "purescript-prelude" : "^4.0.0"
+  },
+  "resolutions": {
+    "purescript-react": "exports"
```
[the remark this answers is not in the document]
Comment (np) [reply bower.json]: This is not needed anymore, thanks.

File: src/Thermite.purs (hunk 1 above)
Comment (np) [reply src/Thermite.purs]: Done.

File: src/Thermite.purs (hunk 5 above)
Comment (np) [reply src/Thermite.purs]: I'm not sure about the formatting issue.

`purescript-react` has a type but here I wanted to keep something simple.

File: src/Thermite.purs (hunk 6 above)
Comment (np) [reply src/Thermite.purs]: There is a single internal use of this function. This might be done differently. As I understand it, children was treated specifically and is now a normal prop. My goal was to make the least amount of changes to the Thermite interface and thus I kept the children outside of props as it was before.

File: src/Thermite.purs (hunk 2 above)
Comment (np) [reply src/Thermite.purs]: This was a left over of messing up too much with the code. Fixed.

File: yarn.lock (hunk 3 above)
Comment (np) [reply yarn.lock]: It's your call but that is the kind of file I prefer to commit as it ensures the devs are using the exact same versions.

File: src/Thermite.purs (hunk 4 above)
Comment (np) [reply src/Thermite.purs]: Ok

File: src/Thermite.purs (hunk 5 above)
Comment (paf31) [reply src/Thermite.purs]: I think it would be better to follow `purescript-react` and just have children be another prop. Is that possible/simple?

File: yarn.lock (hunk 3 above)
Comment (paf31) [reply yarn.lock]: Doesn't pinning versions in package.json ensure that they are at least the same up to minor revisions?

Comment (np) [reply yarn.lock]: Indeed, still there is room for discrepancies between versions, including when testing with CI.
The file is indeed autogenerated and should not be edited, it does not mean it cannot be in the tree.
But again in your fork this is your call.

**Comment by athanclark** [comment]
@np I've taken the liberty of uploading my own version that is stable for purs v0.12.x and react v16.x. Sorry for the delay, and this wasted effort!
